src/Runtime/LoopIterator.php

```diff
@@ -35,33 +37,38 @@ public function __construct($seq)
 
     public function current(): mixed
     {
-        return $this->seq->current();
+        return $this->current['value'];
     }
 
     public function key(): mixed
     {
-        return $this->seq->key();
+        return $this->current['key'];
     }
 
     public function next(): void
     {
-        if ($this->peek) {
-            $this->peek = false;
+        $this->previous = $this->current;
+        if ($this->next) {
+            $this->next = [];
```

Comment:
should we represent that case as `null` instead of an empty array ? This would allow describing the type more precisely by using an array shape, by having the type be `array{valid: bool, key: mixed, value: mixed}|null`, which would help for static analysis of the code.

Reply:
Good idea